File: backend/app/api/v1/ews.py
import datetime as dt
from typing import List, Literal, Optional

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel, Field

from app.engine import views, weather
from app.engine.scenarios import DEFAULT_SCENARIO, SCENARIOS
from app.engine.service import service
from app.engine.villages import BY_ID, INDEX

router = APIRouter(prefix="/ews", tags=["NeerNetra Early Warning Engine"])

InterventionType = Literal["chlorination", "boil_advisory", "medical_camp", "ors_distribution", "source_testing", "awareness_drive"]
InterventionStatus = Literal["dispatched", "in_progress", "completed"]


class ResetIn(BaseModel):
    scenario: Optional[str] = None


class StepIn(BaseModel):
    days: int = Field(1, ge=1, le=14)


class InjectIn(BaseModel):
    type: Literal["heavy_rain", "pipe_burst", "sewage_overflow"]
    villageId: str


class InterventionIn(BaseModel):
    villageId: str
    type: InterventionType
    assignedTo: Optional[str] = None
    notes: Optional[str] = None


class InterventionPatch(BaseModel):
    status: InterventionStatus
    verification: Optional[str] = None


class SymptomsIn(BaseModel):
    diarrhoea: int = Field(0, ge=0)
    vomiting: int = Field(0, ge=0)
    fever: int = Field(0, ge=0)
    jaundice: int = Field(0, ge=0)
    bloodyStool: int = Field(0, ge=0)


class CaseReportIn(BaseModel):
    villageId: str
    reporterRole: Literal["asha", "citizen", "clinic"]
    reporterName: Optional[str] = None
    channel: Literal["app", "offline_sync"]
    date: Optional[dt.date] = None
    symptoms: SymptomsIn
    notes: Optional[str] = None
    clientId: Optional[str] = None


class ReportBatchIn(BaseModel):
    reports: List[CaseReportIn]


class WaterTestIn(BaseModel):
    villageId: str
    source: Literal["sensor", "h2s_kit", "field_kit", "citizen"]
    date: Optional[dt.date] = None
    h2sPositive: Optional[bool] = None
    turbidity: Optional[float] = Field(None, ge=0)
    ph: Optional[float] = Field(None, ge=0, le=14)
    appearance: Optional[Literal["clear", "cloudy", "muddy"]] = None
    notes: Optional[str] = None
    clientId: Optional[str] = None


def _village(vid: str) -> int:
    if vid not in INDEX:
        raise HTTPException(status_code=404, detail=f"Unknown village '{vid}'.")
    return INDEX[vid]


def _dated(payload: BaseModel) -> dict:
    world = service.world
    _village(payload.villageId)
    data = payload.model_dump()
    data["date"] = data["date"] or world.today_date
    if data["date"] > world.today_date:
        raise HTTPException(status_code=400, detail=f"Date {data['date']} is after the current sim date {world.today_date}.")
    if data["date"] < world.book.start:
        raise HTTPException(status_code=400, detail=f"Date {data['date']} is before the simulation history starts ({world.book.start}).")
    return data


def _find(items: list[dict], item_id: int, kind: str) -> dict:
    found = next((x for x in items if x["id"] == item_id), None)
    if found is None:
        raise HTTPException(status_code=404, detail=f"Unknown {kind} {item_id}.")
    return found


@router.get("/state")
def get_state():
    with service.lock:
        return views.sim_state(service.world)


@router.get("/scenarios")
def get_scenarios():
    return [{"id": s.id, "label": s.label, "description": s.description} for s in SCENARIOS.values()]


@router.post("/sim/reset")
def reset_sim(body: Optional[ResetIn] = None):
    scenario = (body.scenario if body else None) or DEFAULT_SCENARIO
    if scenario not in SCENARIOS:
        raise HTTPException(status_code=400, detail=f"Unknown scenario '{scenario}'. Choose one of: {', '.join(SCENARIOS)}.")
    with service.lock:
        return views.sim_state(service.reset(scenario))


@router.post("/sim/step")
def step_sim(body: Optional[StepIn] = None):
    with service.lock:
        service.world.step(service.model, body.days if body else 1)
        service.save()
        return views.sim_state(service.world)


@router.post("/sim/inject")
def inject_event(body: InjectIn):
    _village(body.villageId)
    with service.lock:
        service.world.inject(body.type, body.villageId)
        service.save()
        return views.sim_state(service.world)


@router.get("/villages")
def get_villages():
    with service.lock:
        return views.village_summaries(service.world)


@router.get("/villages/{village_id}")
def get_village(village_id: str):
    i = _village(village_id)
    with service.lock:
        return views.village_detail(service.world, i, service.model.threshold)


@router.get("/alerts")
def get_alerts(status: Literal["active", "all"] = "active"):
    with service.lock:
        alerts = [a for a in service.world.alerts if status == "all" or a["status"] != "resolved"]
        return sorted(alerts, key=lambda a: (a["updatedDate"], a["id"]), reverse=True)


@router.post("/alerts/{alert_id}/ack")
def ack_alert(alert_id: int):
    with service.lock:
        alert = service.world.ack_alert(_find(service.world.alerts, alert_id, "alert"))
        service.save()
        return alert


@router.post("/alerts/{alert_id}/resolve")
def resolve_alert(alert_id: int):
    with service.lock:
        alert = service.world.resolve_alert(_find(service.world.alerts, alert_id, "alert"))
        service.save()
        return alert


@router.get("/notifications")
def get_notifications(
    limit: int = Query(50, ge=1, le=1000),
    role: Optional[Literal["ASHA", "PHC", "DHO", "Citizens"]] = None,
    villageId: Optional[str] = None,
):
    with service.lock:
        rows = [n for n in service.world.notifications
                if (role is None or n["role"] == role) and (villageId is None or n["villageId"] == villageId)]
        return sorted(rows, key=lambda n: n["id"], reverse=True)[:limit]


@router.get("/interventions")
def get_interventions(villageId: Optional[str] = None):
    with service.lock:
        rows = [iv for iv in service.world.interventions if villageId is None or iv["villageId"] == villageId]
        return [views.public(iv) for iv in sorted(rows, key=lambda iv: iv["id"], reverse=True)]


@router.post("/interventions")
def create_intervention(body: InterventionIn):
    _village(body.villageId)
    with service.lock:
        iv = service.world.create_intervention(service.model, body.villageId, body.type, body.assignedTo, body.notes)
        service.save()
        return views.public(iv)


@router.patch("/interventions/{intervention_id}")
def update_intervention(intervention_id: int, body: InterventionPatch):
    with service.lock:
        iv = _find(service.world.interventions, intervention_id, "intervention")
        iv = service.world.update_intervention(service.model, iv, body.status, body.verification)
        service.save()
        return views.public(iv)


@router.post("/reports")
def submit_report(body: CaseReportIn):
    with service.lock:
        report, _ = service.world.add_report(service.model, _dated(body))
        service.save()
        return report


@router.post("/reports/batch")
def submit_reports(body: ReportBatchIn):
    with service.lock:
        rows = [_dated(r) for r in body.reports]
        created = [service.world.add_report(service.model, r)[1] for r in rows]
        service.save()
        return {"accepted": sum(created), "duplicates": len(created) - sum(created)}


@router.get("/reports")
def get_reports(villageId: Optional[str] = None, limit: int = Query(50, ge=1, le=1000)):
    with service.lock:
        rows = [r for r in service.world.reports if villageId is None or r["villageId"] == villageId]
        return sorted(rows, key=lambda r: (r["date"], r["id"]), reverse=True)[:limit]


@router.post("/water-tests")
def submit_water_test(body: WaterTestIn):
    with service.lock:
        test, _ = service.world.add_water_test(service.model, _dated(body))
        service.save()
        return test


@router.get("/advisory/{village_id}")
def get_advisory(village_id: str, lang: Literal["en", "hi", "as"] = "en"):
    i = _village(village_id)
    with service.lock:
        return views.advisory(service.world, i, lang)


@router.get("/model-card")
def get_model_card():
    return service.card


@router.get("/weather/{village_id}")
def get_weather(village_id: str):
    _village(village_id)
    return weather.forecast(BY_ID[village_id])

File: frontend/src/lib/ews.ts
// NeerNetra early-warning engine: typed client for /api/v1/ews/*.
// These shapes are the contract with backend/app/api/v1/ews.py (see docs/ews-api.md).
// JSON is camelCase on the wire. Dates are ISO "yyyy-mm-dd" in simulated time.

export type Tier = 'normal' | 'watch' | 'warning' | 'outbreak';
export type AlertTier = Exclude<Tier, 'normal'>;
export type Lang = 'en' | 'hi' | 'as';
export type WaterSource = 'river' | 'tubewell' | 'spring' | 'pond' | 'piped';
export type AdvisoryStatus = 'safe' | 'boil' | 'do_not_drink';
export type InterventionType =
  | 'chlorination'
  | 'boil_advisory'
  | 'medical_camp'
  | 'ors_distribution'
  | 'source_testing'
  | 'awareness_drive';
export type InterventionStatus = 'dispatched' | 'in_progress' | 'completed';
export type AlertStatus = 'active' | 'acknowledged' | 'resolved';
export type InjectType = 'heavy_rain' | 'pipe_burst' | 'sewage_overflow';
export type ReporterRole = 'asha' | 'citizen' | 'clinic';
export type ReportChannel = 'app' | 'offline_sync';
export type WaterTestSource = 'sensor' | 'h2s_kit' | 'field_kit' | 'citizen';

export interface Kpis {
  villagesAtRisk: number; // villages at tier warning or outbreak
  activeAlerts: number;
  cases7d: number;
  positiveWaterTests7d: number;
  populationCovered: number;
}

export interface SimState {
  currentDate: string;
  scenario: string;
  scenarioLabel: string;
  dayOfScenario: number; // 0 right after reset
  version: number; // bumps on every mutation (step, report, intervention, ack...) so clients know to refetch
  narration: string | null; // one-line story beat for the demo
  kpis: Kpis;
}

export interface Scenario {
  id: string;
  label: string;
  description: string;
}

export interface EarsScores {
  c1: number;
  c2: number;
  c3: number;
  flagged: boolean;
}

export interface LatestReadings {
  turbidity: number | null; // NTU
  ph: number | null;
  tds: number | null; // mg/L
  h2sPositive: boolean | null; // most recent H2S strip result within 7 days
  rainfall24h: number; // mm
  cases24h: number;
  cases7d: number;
}

export interface VillageSummary {
  id: string; // short village code, e.g. "KMR04"
  name: string;
  district: string;
  state: string;
  lat: number;
  lon: number;
  population: number;
  language: Lang; // primary community language for citizen messages
  waterSource: WaterSource;
  catchment: string | null;
  upstreamId: string | null; // village immediately upstream on the same river
  hasSensor: boolean;
  risk: number; // 0..1, ML probability of an outbreak in the next 7 days
  riskDelta: number; // change vs previous sim day
  tier: Tier;
  rank: number; // 1 = highest priority
  ears: EarsScores;
  latest: LatestReadings;
  advisory: AdvisoryStatus;
}

export interface DayPoint {
  date: string;
  rainfall: number;
  turbidity: number | null;
  ph: number | null;
  h2sPositive: boolean | null;
  cases: number;
  risk: number;
  earsC2: number;
  tier: Tier;
}

export interface ExplanationFactor {
  key: string; // e.g. "turbidity", "rainfall", "h2s", "cases", "upstream", "season", "source"
  label: string; // human label, e.g. "Water turbidity"
  value: string; // human value, e.g. "24.5 NTU (5x safe limit)"
  contribution: number; // signed TreeSHAP contribution in log-odds
  share: number; // |contribution| / sum(|contributions|), 0..1
}

export interface Explanation {
  risk: number;
  baseRisk: number; // model expected value as a probability
  summary: string; // one plain-language sentence
  factors: ExplanationFactor[]; // grouped, sorted by |contribution| desc
  detectors: {
    ml: { risk: number; threshold: number; triggered: boolean };
    ears: { c2: number; c3: number; triggered: boolean };
  };
}

export interface RecommendedAction {
  type: InterventionType;
  title: string;
  reason: string;
  priority: 'high' | 'medium' | 'low';
  alreadyActive: boolean;
}

export interface Intervention {
  id: number;
  villageId: string;
  villageName: string;
  type: InterventionType;
  status: InterventionStatus;
  createdDate: string;
  updatedDate: string;
  assignedTo: string;
  notes: string | null;
  verification: string | null; // e.g. "Residual chlorine 0.8 mg/L"
}

export interface WaterTestInput {
  villageId: string;
  source: WaterTestSource;
  date?: string; // defaults to current sim date
  h2sPositive?: boolean | null;
  turbidity?: number | null;
  ph?: number | null;
  appearance?: 'clear' | 'cloudy' | 'muddy' | null;
  notes?: string;
  clientId?: string; // offline de-duplication
}

export interface WaterTest {
  id: number;
  villageId: string;
  date: string;
  source: WaterTestSource;
  h2sPositive: boolean | null;
  turbidity: number | null;
  ph: number | null;
  appearance: 'clear' | 'cloudy' | 'muddy' | null;
  notes: string | null;
}

export interface Symptoms {
  diarrhoea: number;
  vomiting: number;
  fever: number;
  jaundice: number;
  bloodyStool: number;
}

export interface CaseReportInput {
  villageId: string;
  reporterRole: ReporterRole;
  reporterName?: string;
  channel: ReportChannel;
  date?: string; // defaults to current sim date
  symptoms: Symptoms;
  notes?: string;
  clientId?: string; // offline de-duplication
}

export interface CaseReport {
  id: number;
  villageId: string;
  reporterRole: ReporterRole;
  reporterName: string | null;
  channel: ReportChannel;
  date: string;
  symptoms: Symptoms;
  total: number;
  notes: string | null;
}

export interface NeighbourRef {
  id: string;
  name: string;
  tier: Tier;
  risk: number;
}

export interface VillageDetail extends VillageSummary {
  timeseries: DayPoint[]; // last 30 sim days incl. today, ascending
  forecast: { date: string; rainfallMm: number }[]; // next 3 days' rainfall forecast the model used
  explanation: Explanation;
  actions: RecommendedAction[];
  interventions: Intervention[];
  waterTests: WaterTest[]; // newest first, max 20
  reports: CaseReport[]; // newest first, max 20
  upstream: NeighbourRef | null;
  downstream: NeighbourRef[];
}

export interface Alert {
  id: number;
  villageId: string;
  villageName: string;
  tier: AlertTier;
  status: AlertStatus;
  createdDate: string;
  updatedDate: string;
  title: string;
  message: string;
  triggers: string[]; // e.g. ["ML risk 0.72 >= 0.60", "EARS C2 = 3.4"]
  audience: string[]; // e.g. ["ASHA", "PHC", "Citizens"]
}

export type NotificationRole = 'ASHA' | 'PHC' | 'DHO' | 'Citizens';

// No SMS/IVR (no DLT registration): alerts reach people as app notifications only.
export interface OutboundNotification {
  id: number;
  date: string;
  channel: 'push' | 'in_app'; // push = shown as a device notification by the field app; in_app = dashboard inbox only
  role: NotificationRole;
  to: string; // e.g. "ASHA Rina Das (NeerNetra app)" or "Citizens of Chaygaon (NeerNetra app)"
  language: Lang;
  priority: 'normal' | 'high';
  title: string;
  body: string;
  villageId: string;
  alertId: number | null;
}

export interface Advisory {
  villageId: string;
  villageName: string;
  status: AdvisoryStatus;
  tier: Tier;
  headline: string;
  instructions: string[];
  updatedDate: string;
  language: Lang;
}

export interface MethodMetrics {
  method: string;
  recall: number;
  precision: number;
  falseAlarmsPer100VillageDays: number;
  medianLeadDays: number | null; // positive = warned before outbreak onset
  detectedEpisodes: number;
}

export interface ModelCard {
  model: string;
  trainPeriod: string;
  testPeriod: string;
  villages: number;
  trainRows: number;
  testRows: number;
  outbreakDefinition: string;
  syntheticDisclaimer: string;
  auc: number;
  threshold: number;
  testEpisodes: number;
  methods: MethodMetrics[];
  leadTimeHistogram: { leadDays: number; ml: number; ears: number }[];
  featureImportance: { feature: string; label: string; importance: number }[]; // mean |SHAP|, sums to 1
}

export interface WeatherForecast {
  villageId: string;
  source: 'open-meteo' | 'unavailable';
  days: { date: string; precipitationMm: number; precipitationProbability: number | null }[];
}

const BASE = '/api/v1/ews';

type Listener = () => void;
const listeners = new Set<Listener>();

/** Subscribe to local mutations (step, report, ack...) so views can refetch immediately. */
export function onEwsChange(cb: Listener): () => void {
  listeners.add(cb);
  return () => listeners.delete(cb);
}

function emitChange() {
  listeners.forEach((cb) => cb());
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`${BASE}${path}`, {
    ...init,
    headers: { 'Content-Type': 'application/json', ...(init?.headers ?? {}) },
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(`${init?.method ?? 'GET'} ${path} failed (${res.status}): ${body}`);
  }
  return res.json() as Promise<T>;
}

async function mutate<T>(path: string, method: 'POST' | 'PATCH', body?: unknown): Promise<T> {
  const result = await request<T>(path, { method, body: body === undefined ? undefined : JSON.stringify(body) });
  emitChange();
  return result;
}

export const ews = {
  state: () => request<SimState>('/state'),
  scenarios: () => request<Scenario[]>('/scenarios'),
  reset: (scenario?: string) => mutate<SimState>('/sim/reset', 'POST', { scenario }),
  step: (days = 1) => mutate<SimState>('/sim/step', 'POST', { days }),
  inject: (type: InjectType, villageId: string) => mutate<SimState>('/sim/inject', 'POST', { type, villageId }),

  villages: () => request<VillageSummary[]>('/villages'),
  village: (id: string) => request<VillageDetail>(`/villages/${encodeURIComponent(id)}`),

  alerts: (status: 'active' | 'all' = 'active') => request<Alert[]>(`/alerts?status=${status}`),
  ackAlert: (id: number) => mutate<Alert>(`/alerts/${id}/ack`, 'POST'),
  resolveAlert: (id: number) => mutate<Alert>(`/alerts/${id}/resolve`, 'POST'),
  notifications: (opts: { limit?: number; role?: NotificationRole; villageId?: string } = {}) => {
    const q = new URLSearchParams({ limit: String(opts.limit ?? 50) });
    if (opts.role) q.set('role', opts.role);
    if (opts.villageId) q.set('villageId', opts.villageId);
    return request<OutboundNotification[]>(`/notifications?${q}`);
  },

  interventions: (villageId?: string) =>
    request<Intervention[]>(`/interventions${villageId ? `?villageId=${encodeURIComponent(villageId)}` : ''}`),
  createIntervention: (input: { villageId: string; type: InterventionType; assignedTo?: string; notes?: string }) =>
    mutate<Intervention>('/interventions', 'POST', input),
  updateIntervention: (id: number, input: { status: InterventionStatus; verification?: string }) =>
    mutate<Intervention>(`/interventions/${id}`, 'PATCH', input),

  submitReport: (input: CaseReportInput) => mutate<CaseReport>('/reports', 'POST', input),
  submitReports: (reports: CaseReportInput[]) =>
    mutate<{ accepted: number; duplicates: number }>('/reports/batch', 'POST', { reports }),
  reports: (villageId?: string, limit = 50) =>
    request<CaseReport[]>(`/reports?limit=${limit}${villageId ? `&villageId=${encodeURIComponent(villageId)}` : ''}`),
  submitWaterTest: (input: WaterTestInput) => mutate<WaterTest>('/water-tests', 'POST', input),

  advisory: (villageId: string, lang: Lang = 'en') =>
    request<Advisory>(`/advisory/${encodeURIComponent(villageId)}?lang=${lang}`),
  modelCard: () => request<ModelCard>('/model-card'),
  weather: (villageId: string) => request<WeatherForecast>(`/weather/${encodeURIComponent(villageId)}`),
};

export const TIER_ORDER: Record<Tier, number> = { normal: 0, watch: 1, warning: 2, outbreak: 3 };
